File: testdata/cpp/simple.cpp
#include <iostream>
#include "sensor.h"

class SensorManager {
public:
    void initialize();
    double readTemperature();
private:
    int sensorId;
};

void SensorManager::initialize() {
    calibrate();
}

double SensorManager::readTemperature() {
    return getReading(sensorId);
}

struct SensorConfig {
    int pin;
    double threshold;
};

enum SensorType {
    TEMPERATURE,
    HUMIDITY,
    PRESSURE
};

void setup() {
    SensorManager mgr;
    mgr.initialize();
    double temp = mgr.readTemperature();
    logValue(temp);
    // nw-434: calls with an explicit template-argument list between the
    // callee name and the argument list. tree-sitter-cpp folds these into a
    // distinct `template_function`/`template_method` callee node instead of
    // the plain `identifier`/`qualified_identifier`/`field_expression` a
    // same-shaped call without `<...>` would use, so a query that only
    // handles the untemplated forms drops the call entirely.
    logGeneric<double>(temp);           // free function
    sensors::logGeneric<double>(temp);  // qualified (namespace- or class-scoped)
    mgr.logReading<int>(temp);          // member call (`obj.f<T>(...)`)
}

File: testdata/zig/simple.zig
const std = @import("std");
const math = @import("math.zig");

pub const SensorKind = enum {
    temperature,
    humidity,
    pressure,
};

pub const SensorConfig = struct {
    name: []const u8,
    kind: SensorKind,
    threshold: f64,
};

const InternalState = union {
    active: bool,
    error_code: u32,
};

pub fn initialize(config: SensorConfig) void {
    std.debug.print("Initializing sensor: {s}\n", .{config.name});
    calibrate(config);
}

fn calibrate(config: SensorConfig) void {
    _ = config;
}

pub fn main() !void {
    const config = SensorConfig{
        .name = "temp-1",
        .kind = .temperature,
        .threshold = 100.0,
    };
    initialize(config);
}
